GitTfs/Core/GitRepository.cs

```diff
@@ -520,18 +528,23 @@ public bool CreateBranch(string gitBranchName, string target)
             return reference != null;
         }
 
-        private readonly Dictionary<long, string> changesetsCache = new Dictionary<long, string>();
+        private readonly Dictionary<long, IList<string>> changesetsCache = new Dictionary<long, IList<string>>();
         private bool cacheIsFull = false;
 
         public string FindCommitHashByChangesetId(long changesetId, string tfsPath)
         {
-            var commit = FindCommitByChangesetId(changesetId, tfsPath);
+            var commit = FindCommitsByChangesetId(changesetId).FirstOrDefault(c => c.Message.Contains(tfsPath));
```

Comment:
I don't like a lot this solution which consist in travelling along a lot of commit to construct a list and then filter this list with the tfs path :( Because that's not very good in terms of performance...

I'm pretty shure that we can write a private method that take an optional path and :

* if the path is not empty, return immediatly if the commit with the good id and path is found 
* if the path is empty, continue to look for this specific changeset id through all the commits

This method could be wrapped in 2 public methods one which accept a tfs path and one another without...